Feature: Add \x hex escape sequence support in strings

## Description

Add support for `\x` hex escape sequences in string literals. This allows embedding arbitrary bytes using hexadecimal notation, e.g., `"\x1b"` for the ASCII escape character (27).

## Motivation

ANSI terminal color codes require the escape character (`\x1b` or ASCII 27). Without hex escape support, it is impossible to write terminal styling utilities in pure EZ.

Example use case:
```ez
const RED = "\x1b[31m"
const RESET = "\x1b[0m"

do red(text string) -> string {
    return "${RED}${text}${RESET}"
}
```

## Current Behavior

```
error[E1006]: invalid escape sequence '\x' in string
```

## Proposed Syntax

- `\xNN` where NN is exactly two hexadecimal digits (0-9, a-f, A-F)
- Examples: `\x1b`, `\x00`, `\xFF`, `\x0a`

## Implementation

Requires changes to the lexer to recognize and process `\x` escape sequences.

--- a/integration-tests/fail/errors/E1006_invalid_hex_escape.ez
+++ b/integration-tests/fail/errors/E1006_invalid_hex_escape.ez
@@ -1,0 +1,8 @@
+/*
+ * Error Test: E1006 - invalid hex escape
+ * Expected: "hex" or "invalid"
+ */
+
+do main() {
+    temp bad = "\xGG"  // GG is not valid hex
+}

--- a/integration-tests/pass/core/hex-escape.ez
+++ b/integration-tests/pass/core/hex-escape.ez
@@ -1,0 +1,102 @@
+/*
+ * hex-escape.ez - Test hex escape sequences in strings and chars
+ */
+
+import @std
+using std
+
+do main() {
+    println("Testing hex escape sequences...")
+    temp passed = 0
+    temp failed = 0
+
+    // Test 1: Hex escape for null byte
+    temp null_byte = "\x00"
+    if len(null_byte) == 1 {
+        println("  [PASS] \\x00 produces single byte")
+        passed += 1
+    } otherwise {
+        println("  [FAIL] \\x00 length incorrect")
+        failed += 1
+    }
+
+    // Test 2: Hex escape for newline (0x0a = 10)
+    temp newline = "\x0a"
+    if newline == "\n" {
+        println("  [PASS] \\x0a equals \\n")
+        passed += 1
+    } otherwise {
+        println("  [FAIL] \\x0a does not equal \\n")
+        failed += 1
+    }
+
+    // Test 3: Hex escape for tab (0x09 = 9)
+    temp tab = "\x09"
+    if tab == "\t" {
+        println("  [PASS] \\x09 equals \\t")
+        passed += 1
+    } otherwise {
+        println("  [FAIL] \\x09 does not equal \\t")
+        failed += 1
+    }
+
+    // Test 4: Hex escape for carriage return (0x0d = 13)
+    temp cr = "\x0d"
+    if cr == "\r" {
+        println("  [PASS] \\x0d equals \\r")
+        passed += 1
+    } otherwise {
+        println("  [FAIL] \\x0d does not equal \\r")
+        failed += 1
+    }
+
+    // Test 5: Uppercase hex digits
+    temp upper = "\x1B"
+    temp lower = "\x1b"
+    if upper == lower {
+        println("  [PASS] \\x1B equals \\x1b (case insensitive)")
+        passed += 1
+    } otherwise {
+        println("  [FAIL] hex digits should be case insensitive")
+        failed += 1
+    }
+
+    // Test 6: Multiple hex escapes in one string
+    temp multi = "\x48\x69"
+    if multi == "Hi" {
+        println("  [PASS] \\x48\\x69 equals 'Hi'")
+        passed += 1
+    } otherwise {
+        println("  [FAIL] \\x48\\x69 should equal 'Hi', got '${multi}'")
+        failed += 1
+    }
+
+    // Test 7: Hex escape mixed with regular text
+    temp mixed = "A\x42C"
+    if mixed == "ABC" {
+        println("  [PASS] 'A\\x42C' equals 'ABC'")
+        passed += 1
+    } otherwise {
+        println("  [FAIL] mixed hex/text failed")
+        failed += 1
+    }
+
+    // Test 8: ANSI escape sequence (ESC = 0x1b = 27)
+    temp esc = "\x1b"
+    if len(esc) == 1 {
+        println("  [PASS] ANSI escape character length is 1")
+        passed += 1
+    } otherwise {
+        println("  [FAIL] ANSI escape character length incorrect")
+        failed += 1
+    }
+
+    // Summary
+    println("")
+    println("Passed: ${passed}")
+    println("Failed: ${failed}")
+
+    if failed > 0 {
+        panic("Some tests failed!")
+    }
+}

--- a/pkg/lexer/lexer.go
+++ b/pkg/lexer/lexer.go
@@ -627,12 +627,24 @@ func (l *Lexer) readString() (string, bool) {
 			// Check for valid escape sequences
 			next := l.peekChar()
 			validEscapes := map[byte]bool{
-				'n': true, 't': true, 'r': true, '\\': true, '"': true, '\'': true, '0': true,
+				'n': true, 't': true, 'r': true, '\\': true, '"': true, '\'': true, '0': true, 'x': true,
 			}
 			if !validEscapes[next] && next != 0 {
 				l.addError("E1006", "invalid escape sequence '\\"+string(next)+"' in string", l.line, l.column)
 			}
 			l.readChar() // skip escaped character
+
+			// For hex escapes, validate and skip the two hex digits
+			if next == 'x' {
+				for j := 0; j < 2; j++ {
+					hexChar := l.peekChar()
+					if !isHexDigit(hexChar) {
+						l.addError("E1006", "invalid hex escape sequence: expected two hex digits after \\x", l.line, l.column)
+						break
+					}
+					l.readChar()
+				}
+			}
 		}
 	}
 }
@@ -692,12 +704,26 @@ func (l *Lexer) readCharValue() string {
 		l.readChar()
 		// Validate escape sequence
 		validEscapes := map[byte]bool{
-			'n': true, 't': true, 'r': true, '\\': true, '"': true, '\'': true, '0': true,
+			'n': true, 't': true, 'r': true, '\\': true, '"': true, '\'': true, '0': true, 'x': true,
 		}
 		if !validEscapes[l.ch] && l.ch != 0 {
 			l.addError("E1007", "invalid escape sequence '\\"+string(l.ch)+"' in character literal", startLine, startColumn)
 		}
-		ch = "\\" + string(l.ch)
+
+		// For hex escapes, read the two hex digits
+		if l.ch == 'x' {
+			ch = "\\x"
+			for j := 0; j < 2; j++ {
+				l.readChar()
+				if !isHexDigit(l.ch) {
+					l.addError("E1007", "invalid hex escape sequence: expected two hex digits after \\x", startLine, startColumn)
+					break
+				}
+				ch += string(l.ch)
+			}
+		} else {
+			ch = "\\" + string(l.ch)
+		}
 	}
 
 	l.readChar() // move to what should be closing quote

--- a/pkg/parser/parser.go
+++ b/pkg/parser/parser.go
@@ -2809,6 +2809,21 @@ func processEscapeSequences(s string) string {
 			case '0':
 				result.WriteByte(0)
 				i++
+			case 'x':
+				// Hex escape: \xNN where NN are two hex digits
+				if i+3 < len(s) {
+					hexStr := s[i+2 : i+4]
+					if val, err := strconv.ParseUint(hexStr, 16, 8); err == nil {
+						result.WriteByte(byte(val))
+						i += 3 // skip \xNN (we already skip one with the loop increment)
+					} else {
+						// Invalid hex, keep as-is (lexer already reported error)
+						result.WriteByte(s[i])
+					}
+				} else {
+					// Not enough characters, keep as-is
+					result.WriteByte(s[i])
+				}
 			default:
 				// Not a valid escape, keep as-is (lexer already reported error)
 				result.WriteByte(s[i])
@@ -2946,6 +2961,18 @@ func (p *Parser) parseCharValue() Expression {
 			ch = '\''
 		case '0':
 			ch = 0
+		case 'x':
+			// Hex escape: \xNN where NN are two hex digits
+			if len(literal) >= 4 {
+				hexStr := literal[2:4]
+				if val, err := strconv.ParseUint(hexStr, 16, 8); err == nil {
+					ch = rune(val)
+				} else {
+					ch = rune(literal[1])
+				}
+			} else {
+				ch = rune(literal[1])
+			}
 		default:
 			ch = rune(literal[1])
 		}

